File: lib/utils/date_optimizer.dart
import '../models/holiday.dart';
import '../models/best_vacation_period.dart';
import '../models/school_holiday_period.dart';

enum SchoolHolidayPreference { neutral, favor, avoid }

class DateOptimizer {
  static List<BestVacationPeriod> findAllOptimizedPeriods({
    required List<Holiday> holidays,
    required int vacationDaysToUse,
    required int availableRttDays,
    required int year,
    required int month,
    List<SchoolHolidayPeriod> schoolHolidayPeriods = const [],
    SchoolHolidayPreference schoolHolidayPreference =
        SchoolHolidayPreference.neutral,
    bool allowSchoolHolidayOverlap = false,
  }) {
    return _findCandidatePeriods(
      holidays: holidays,
      vacationDaysToUse: vacationDaysToUse,
      availableRttDays: availableRttDays,
      year: year,
      month: month,
      requireExactVacationDays: true,
      schoolHolidayPeriods: schoolHolidayPeriods,
      schoolHolidayPreference: schoolHolidayPreference,
      allowSchoolHolidayOverlap: allowSchoolHolidayOverlap,
    );
  }

  static List<BestVacationPeriod> findBestDistributedPeriods({
    required List<Holiday> holidays,
    required int vacationDaysToUse,
    required int availableRttDays,
    required int year,
    required int month,
    List<SchoolHolidayPeriod> schoolHolidayPeriods = const [],
    SchoolHolidayPreference schoolHolidayPreference =
        SchoolHolidayPreference.neutral,
    bool allowSchoolHolidayOverlap = false,
  }) {
    final candidates = _findCandidatePeriods(
      holidays: holidays,
      vacationDaysToUse: vacationDaysToUse,
      availableRttDays: availableRttDays,
      year: year,
      month: month,
      requireExactVacationDays: false,
      schoolHolidayPeriods: schoolHolidayPeriods,
      schoolHolidayPreference: schoolHolidayPreference,
      allowSchoolHolidayOverlap: allowSchoolHolidayOverlap,
    ).where((period) => period.vacationDaysUsed > 0).toList();

    if (candidates.isEmpty) {
      return [];
    }

    candidates.sort((a, b) {
      final endCompare = a.endDate.compareTo(b.endDate);
      if (endCompare != 0) {
        return endCompare;
      }
      return a.startDate.compareTo(b.startDate);
    });

    final previousNonOverlapping = List<int>.filled(candidates.length, -1);
    for (var i = 0; i < candidates.length; i++) {
      for (var j = i - 1; j >= 0; j--) {
        if (!candidates[j].endDate.isAfter(candidates[i].startDate)) {
          previousNonOverlapping[i] = j;
          break;
        }
      }
    }

    final dp = List.generate(
      candidates.length + 1,
      (_) => List<_BundleScore?>.filled(vacationDaysToUse + 1, null),
    );
    final take = List.generate(
      candidates.length + 1,
      (_) => List<bool>.filled(vacationDaysToUse + 1, false),
    );

    for (var budget = 0; budget <= vacationDaysToUse; budget++) {
      dp[0][budget] = const _BundleScore.empty();
    }

    for (var i = 1; i <= candidates.length; i++) {
      final candidate = candidates[i - 1];
      for (var budget = 0; budget <= vacationDaysToUse; budget++) {
        var best = dp[i - 1][budget]!;
          if (candidate.vacationDaysUsed <= budget) {
          final previousIndex = previousNonOverlapping[i - 1] + 1;
            final previousScore = dp[previousIndex][budget - candidate.vacationDaysUsed];
            if (previousScore != null) {
              final withCandidate = previousScore.add(candidate);
            if (withCandidate.isBetterThan(best)) {
              best = withCandidate;
              take[i][budget] = true;
            }
          }
        }
        dp[i][budget] = best;
      }
    }

    final selected = <BestVacationPeriod>[];
    var i = candidates.length;
    var budget = vacationDaysToUse;
    while (i > 0 && budget >= 0) {
      if (take[i][budget]) {
        final candidate = candidates[i - 1];
        selected.add(candidate);
        budget -= candidate.vacationDaysUsed;
        i = previousNonOverlapping[i - 1] + 1;
      } else {
        i -= 1;
      }
    }

    selected.sort((a, b) => a.startDate.compareTo(b.startDate));
    return selected;
  }

  static List<BestVacationPeriod> _findCandidatePeriods({
    required List<Holiday> holidays,
    required int vacationDaysToUse,
    required int availableRttDays,
    required int year,
    required int month,
    required bool requireExactVacationDays,
    required List<SchoolHolidayPeriod> schoolHolidayPeriods,
    required SchoolHolidayPreference schoolHolidayPreference,
    required bool allowSchoolHolidayOverlap,
  }) {
    final List<BestVacationPeriod> results = [];
    final Set<String> seenRanges = <String>{};

    if (vacationDaysToUse <= 0) {
      return results;
    }

    final validHolidays =
        holidays
            .where(
              (holiday) =>
                  holiday.date.year == year &&
                  holiday.date.month == month &&
                  holiday.date.weekday != DateTime.saturday &&
                  holiday.date.weekday != DateTime.sunday,
            )
            .toList()
          ..sort((a, b) => a.date.compareTo(b.date));

    if (validHolidays.isEmpty) {
      return results;
    }

    final monthStart = DateTime(year, month, 1);
    final monthEnd = DateTime(year, month + 1, 0);
    final holidayLookup = <DateTime, Holiday>{
      for (final holiday in validHolidays) _stripTime(holiday.date): holiday,
    };

    for (
      DateTime start = monthStart;
      !start.isAfter(monthEnd);
      start = start.add(const Duration(days: 1))
    ) {
      for (
        DateTime end = start;
        !end.isAfter(monthEnd);
        end = end.add(const Duration(days: 1))
      ) {
        final fullRange = List<DateTime>.generate(
          end.difference(start).inDays + 1,
          (index) => start.add(Duration(days: index)),
        );

        final includedHolidays = fullRange
            .map((date) => holidayLookup[_stripTime(date)])
            .whereType<Holiday>()
            .toList();

        if (includedHolidays.isEmpty) {
          continue;
        }

        final vacationDates = fullRange.where((date) {
          final normalizedDate = _stripTime(date);
          final isWeekday =
              date.weekday >= DateTime.monday &&
              date.weekday <= DateTime.friday;
          return isWeekday && !holidayLookup.containsKey(normalizedDate);
        }).toList();

        final matchesBudget = requireExactVacationDays
            ? vacationDates.length == vacationDaysToUse
            : vacationDates.length <= vacationDaysToUse;
        if (!matchesBudget) {
          continue;
        }

        final rttDaysUsed = vacationDates.length.clamp(0, availableRttDays);
        final paidLeaveDaysUsed = vacationDates.length - rttDaysUsed;

        final weekendDates = fullRange
            .where(
              (date) =>
                  date.weekday == DateTime.saturday ||
                  date.weekday == DateTime.sunday,
            )
            .toList();

        final bridgeDates = vacationDates.where((date) {
          final previousDay = _stripTime(
            date.subtract(const Duration(days: 1)),
          );
          final nextDay = _stripTime(date.add(const Duration(days: 1)));
          return holidayLookup.containsKey(previousDay) ||
              holidayLookup.containsKey(nextDay);
        }).toList();

        final totalDaysOff = fullRange.length;
        final normalizedPaidLeaveDays = paidLeaveDaysUsed.clamp(1, totalDaysOff);
        final worthScore = totalDaysOff / normalizedPaidLeaveDays;
        final schoolHolidayDates = fullRange.where((date) {
          return schoolHolidayPeriods.any((period) => period.contains(date));
        }).toList();
        if (!allowSchoolHolidayOverlap && schoolHolidayDates.isNotEmpty) {
          continue;
        }
        final rankingScore = _computeRankingScore(
          worthScore: worthScore,
          schoolHolidayOverlapDays: schoolHolidayDates.length,
          preference: schoolHolidayPreference,
        );
        final signature =
            '${start.toIso8601String()}_${end.toIso8601String()}_${vacationDates.length}_$rttDaysUsed';

        if (totalDaysOff <= vacationDates.length ||
            seenRanges.contains(signature)) {
          continue;
        }

        final paidLeaveDates = vacationDates.skip(rttDaysUsed).toList();
        final rttDates = vacationDates.take(rttDaysUsed).toList();

        seenRanges.add(signature);
        results.add(
          BestVacationPeriod(
            startDate: start,
            endDate: end,
            vacationDaysUsed: vacationDates.length,
            paidLeaveDaysUsed: paidLeaveDaysUsed,
            rttDaysUsed: rttDaysUsed,
            totalDaysOff: totalDaysOff,
            relatedHoliday: includedHolidays.first,
            includedHolidays: includedHolidays,
            vacationDates: vacationDates,
            paidLeaveDates: paidLeaveDates,
            rttDates: rttDates,
            bridgeDates: bridgeDates,
            weekendDates: weekendDates,
            schoolHolidayDates: schoolHolidayDates,
            holidayDate: includedHolidays.first.date,
            worthScore: worthScore,
            rankingScore: rankingScore,
          ),
        );
      }
    }

    results.sort((a, b) {
      final worthCompare = b.rankingScore.compareTo(a.rankingScore);
      if (worthCompare != 0) {
        return worthCompare;
      }

      final totalDaysCompare = b.totalDaysOff.compareTo(a.totalDaysOff);
      if (totalDaysCompare != 0) {
        return totalDaysCompare;
      }

      return a.startDate.compareTo(b.startDate);
    });
    return results;
  }

  static DateTime _stripTime(DateTime date) {
    return DateTime(date.year, date.month, date.day);
  }

  static double _computeRankingScore({
    required double worthScore,
    required int schoolHolidayOverlapDays,
    required SchoolHolidayPreference preference,
  }) {
    switch (preference) {
      case SchoolHolidayPreference.favor:
        return worthScore + (schoolHolidayOverlapDays * 0.2);
      case SchoolHolidayPreference.avoid:
        return worthScore - (schoolHolidayOverlapDays * 0.2);
      case SchoolHolidayPreference.neutral:
        return worthScore;
    }
  }
}

class _BundleScore {
  const _BundleScore({
    required this.totalDaysOff,
    required this.totalBookedDays,
    required this.periodCount,
    required this.schoolHolidayDays,
  });

  const _BundleScore.empty()
      : totalDaysOff = 0,
        totalBookedDays = 0,
        periodCount = 0,
        schoolHolidayDays = 0;

  final int totalDaysOff;
  final int totalBookedDays;
  final int periodCount;
  final int schoolHolidayDays;

  _BundleScore add(BestVacationPeriod period) {
    return _BundleScore(
      totalDaysOff: totalDaysOff + period.totalDaysOff,
      totalBookedDays: totalBookedDays + period.vacationDaysUsed,
      periodCount: periodCount + 1,
      schoolHolidayDays: schoolHolidayDays + period.schoolHolidayDates.length,
    );
  }

  bool isBetterThan(_BundleScore other) {
    if (totalDaysOff != other.totalDaysOff) {
      return totalDaysOff > other.totalDaysOff;
    }
    if (totalBookedDays != other.totalBookedDays) {
      return totalBookedDays < other.totalBookedDays;
    }
    return periodCount > other.periodCount;
  }
}

File: lib/domain/date-optimizer.ts
import {
  AnnualPlanningStrategy,
  AnnualPlanResult,
  AnnualPlanSegment,
  BestVacationPeriod,
  Holiday,
  SchoolHolidayPeriod,
  SchoolHolidayPreference,
} from "@/lib/domain/types";
import { inclusiveDateRange, isWeekend, stripTime } from "@/lib/domain/date-utils";

type CandidateParams = {
  holidays: Holiday[];
  vacationDaysToUse: number;
  availableRttDays: number;
  year: number;
  month: number;
  requireExactVacationDays: boolean;
  schoolHolidayPeriods?: SchoolHolidayPeriod[];
  schoolHolidayPreference?: SchoolHolidayPreference;
  allowSchoolHolidayOverlap?: boolean;
};

type AnnualPlanningParams = {
  holidays: Holiday[];
  paidLeaveBudget: number;
  availableRttDays: number;
  year: number;
  strategy: AnnualPlanningStrategy;
  schoolHolidayPeriods?: SchoolHolidayPeriod[];
  schoolHolidayPreference?: SchoolHolidayPreference;
  allowSchoolHolidayOverlap?: boolean;
};

export const DateOptimizer = {
  findAllOptimizedPeriods({
    holidays,
    vacationDaysToUse,
    availableRttDays,
    year,
    month,
    schoolHolidayPeriods = [],
    schoolHolidayPreference = "neutral",
    allowSchoolHolidayOverlap = false,
  }: Omit<CandidateParams, "requireExactVacationDays">): BestVacationPeriod[] {
    return findCandidatePeriods({
      holidays,
      vacationDaysToUse,
      availableRttDays,
      year,
      month,
      requireExactVacationDays: true,
      schoolHolidayPeriods,
      schoolHolidayPreference,
      allowSchoolHolidayOverlap,
    });
  },

  findBestDistributedPeriods({
    holidays,
    vacationDaysToUse,
    availableRttDays,
    year,
    month,
    schoolHolidayPeriods = [],
    schoolHolidayPreference = "neutral",
    allowSchoolHolidayOverlap = false,
  }: Omit<CandidateParams, "requireExactVacationDays">): BestVacationPeriod[] {
    const candidates = findCandidatePeriods({
      holidays,
      vacationDaysToUse,
      availableRttDays,
      year,
      month,
      requireExactVacationDays: false,
      schoolHolidayPeriods,
      schoolHolidayPreference,
      allowSchoolHolidayOverlap,
    }).filter((period) => period.vacationDaysUsed > 0);

    if (candidates.length === 0) {
      return [];
    }

    candidates.sort((left, right) => {
      const endCompare = left.endDate.getTime() - right.endDate.getTime();
      if (endCompare !== 0) {
        return endCompare;
      }
      return left.startDate.getTime() - right.startDate.getTime();
    });

    const previousNonOverlapping = new Array<number>(candidates.length).fill(-1);
    for (let i = 0; i < candidates.length; i += 1) {
      for (let j = i - 1; j >= 0; j -= 1) {
        if (candidates[j].endDate.getTime() <= candidates[i].startDate.getTime()) {
          previousNonOverlapping[i] = j;
          break;
        }
      }
    }

    const dp: Array<Array<BundleScore | null>> = Array.from(
      { length: candidates.length + 1 },
      () => Array.from({ length: vacationDaysToUse + 1 }, () => null),
    );
    const take: boolean[][] = Array.from({ length: candidates.length + 1 }, () =>
      Array.from({ length: vacationDaysToUse + 1 }, () => false),
    );

    for (let budget = 0; budget <= vacationDaysToUse; budget += 1) {
      dp[0][budget] = BundleScore.empty();
    }

    for (let i = 1; i <= candidates.length; i += 1) {
      const candidate = candidates[i - 1];
      for (let budget = 0; budget <= vacationDaysToUse; budget += 1) {
        let best = dp[i - 1][budget]!;

        if (candidate.vacationDaysUsed <= budget) {
          const previousIndex = previousNonOverlapping[i - 1] + 1;
          const previousScore = dp[previousIndex][budget - candidate.vacationDaysUsed];
          if (previousScore) {
            const withCandidate = previousScore.add(candidate);
            if (withCandidate.isBetterThan(best)) {
              best = withCandidate;
              take[i][budget] = true;
            }
          }
        }

        dp[i][budget] = best;
      }
    }

    const selected: BestVacationPeriod[] = [];
    let i = candidates.length;
    let budget = vacationDaysToUse;

    while (i > 0 && budget >= 0) {
      if (take[i][budget]) {
        const candidate = candidates[i - 1];
        selected.push(candidate);
        budget -= candidate.vacationDaysUsed;
        i = previousNonOverlapping[i - 1] + 1;
      } else {
        i -= 1;
      }
    }

    return selected.sort((left, right) => left.startDate.getTime() - right.startDate.getTime());
  },

  findFlexiblePeriods({
    holidays,
    vacationDaysToUse,
    availableRttDays,
    year,
    month,
    schoolHolidayPeriods = [],
    schoolHolidayPreference = "neutral",
    allowSchoolHolidayOverlap = false,
  }: Omit<CandidateParams, "requireExactVacationDays">): BestVacationPeriod[] {
    return findCandidatePeriods({
      holidays,
      vacationDaysToUse,
      availableRttDays,
      year,
      month,
      requireExactVacationDays: false,
      schoolHolidayPeriods,
      schoolHolidayPreference,
      allowSchoolHolidayOverlap,
    });
  },

  buildAnnualPlan({
    holidays,
    paidLeaveBudget,
    availableRttDays,
    year,
    strategy,
    schoolHolidayPeriods = [],
    schoolHolidayPreference = "neutral",
    allowSchoolHolidayOverlap = false,
  }: AnnualPlanningParams): AnnualPlanResult {
    if (paidLeaveBudget <= 0) {
      return {
        strategy,
        scope: "yearly",
        segments: [],
        totalPaidLeaveUsed: 0,
        totalRttUsed: 0,
        totalDaysOff: 0,
        remainingBudget: 0,
      };
    }

    const bySignature = new Map<string, AnnualPlanSegment>();

    for (let month = 1; month <= 12; month += 1) {
      const monthCandidates = findCandidatePeriods({
        holidays,
        vacationDaysToUse: paidLeaveBudget,
        availableRttDays,
        year,
        month,
        requireExactVacationDays: false,
        schoolHolidayPeriods,
        schoolHolidayPreference,
        allowSchoolHolidayOverlap,
      }).slice(0, 12);

      for (const period of monthCandidates) {
        const signature = [
          period.startDate.toISOString(),
          period.endDate.toISOString(),
          period.paidLeaveDaysUsed,
          period.rttDaysUsed,
        ].join("_");

        if (!bySignature.has(signature)) {
          bySignature.set(signature, {
            month,
            strategy,
            period,
          });
        }
      }
    }

    const candidates = [...bySignature.values()].sort((left, right) => {
      const scoreCompare =
        scoreAnnualCandidate(right.period, strategy) - scoreAnnualCandidate(left.period, strategy);
      if (scoreCompare !== 0) {
        return scoreCompare;
      }

      const paidLeaveCompare = left.period.paidLeaveDaysUsed - right.period.paidLeaveDaysUsed;
      if (paidLeaveCompare !== 0) {
        return paidLeaveCompare;
      }

      return left.period.startDate.getTime() - right.period.startDate.getTime();
    });

    const selected: AnnualPlanSegment[] = [];
    let remainingBudget = paidLeaveBudget;

    for (const candidate of candidates) {
      if (candidate.period.paidLeaveDaysUsed > remainingBudget) {
        continue;
      }

      if (
        selected.some((segment) => periodsOverlap(segment.period, candidate.period))
      ) {
        continue;
      }

      if (
        strategy === "balanced" &&
        selected.some((segment) => segment.month === candidate.month)
      ) {
        continue;
      }

      selected.push(candidate);
      remainingBudget -= candidate.period.paidLeaveDaysUsed;
    }

    selected.sort(
      (left, right) => left.period.startDate.getTime() - right.period.startDate.getTime(),
    );

    return {
      strategy,
      scope: "yearly",
      segments: selected,
      totalPaidLeaveUsed: selected.reduce(
        (sum, segment) => sum + segment.period.paidLeaveDaysUsed,
        0,
      ),
      totalRttUsed: selected.reduce((sum, segment) => sum + segment.period.rttDaysUsed, 0),
      totalDaysOff: selected.reduce((sum, segment) => sum + segment.period.totalDaysOff, 0),
      remainingBudget,
    };
  },
};

function findCandidatePeriods({
  holidays,
  vacationDaysToUse,
  availableRttDays,
  year,
  month,
  requireExactVacationDays,
  schoolHolidayPeriods = [],
  schoolHolidayPreference = "neutral",
  allowSchoolHolidayOverlap = false,
}: CandidateParams): BestVacationPeriod[] {
  const results: BestVacationPeriod[] = [];
  const seenRanges = new Set<string>();

  if (vacationDaysToUse <= 0) {
    return results;
  }

  const validHolidays = holidays
    .filter((holiday) => {
      const date = holiday.date;
      return (
        date.getFullYear() === year &&
        date.getMonth() + 1 === month &&
        !isWeekend(date)
      );
    })
    .sort((left, right) => left.date.getTime() - right.date.getTime());

  if (validHolidays.length === 0) {
    return results;
  }

  const monthStart = new Date(year, month - 1, 1);
  const monthEnd = new Date(year, month, 0);
  const holidayLookup = new Map<number, Holiday>(
    validHolidays.map((holiday) => [stripTime(holiday.date).getTime(), holiday]),
  );

  for (let start = new Date(monthStart); start <= monthEnd; start.setDate(start.getDate() + 1)) {
    for (let end = new Date(start); end <= monthEnd; end.setDate(end.getDate() + 1)) {
      const fullRange = inclusiveDateRange(new Date(start), new Date(end));
      const includedHolidays = fullRange
        .map((date) => holidayLookup.get(stripTime(date).getTime()))
        .filter(Boolean) as Holiday[];

      if (includedHolidays.length === 0) {
        continue;
      }

      const vacationDates = fullRange.filter((date) => {
        const normalizedDate = stripTime(date);
        const weekday = normalizedDate.getDay();
        const isWeekday = weekday >= 1 && weekday <= 5;
        return isWeekday && !holidayLookup.has(normalizedDate.getTime());
      });

      const matchesBudget = requireExactVacationDays
        ? vacationDates.length === vacationDaysToUse
        : vacationDates.length <= vacationDaysToUse;

      if (!matchesBudget) {
        continue;
      }

      const rttDaysUsed = Math.min(vacationDates.length, availableRttDays);
      const paidLeaveDaysUsed = vacationDates.length - rttDaysUsed;

      const weekendDates = fullRange.filter(isWeekend);
      const bridgeDates = vacationDates.filter((date) => {
        const previousDay = stripTime(new Date(date.getFullYear(), date.getMonth(), date.getDate() - 1));
        const nextDay = stripTime(new Date(date.getFullYear(), date.getMonth(), date.getDate() + 1));
        return holidayLookup.has(previousDay.getTime()) || holidayLookup.has(nextDay.getTime());
      });

      const schoolHolidayDates = fullRange.filter((date) =>
        schoolHolidayPeriods.some((period) => containsSchoolHoliday(period, date)),
      );

      if (!allowSchoolHolidayOverlap && schoolHolidayDates.length > 0) {
        continue;
      }

      const totalDaysOff = fullRange.length;
      const normalizedPaidLeaveDays = clamp(paidLeaveDaysUsed, 1, totalDaysOff);
      const worthScore = totalDaysOff / normalizedPaidLeaveDays;
      const rankingScore = computeRankingScore({
        worthScore,
        schoolHolidayOverlapDays: schoolHolidayDates.length,
        preference: schoolHolidayPreference,
      });

      const signature = [
        stripTime(start).toISOString(),
        stripTime(end).toISOString(),
        vacationDates.length,
        rttDaysUsed,
      ].join("_");

      if (totalDaysOff <= vacationDates.length || seenRanges.has(signature)) {
        continue;
      }

      const paidLeaveDates = vacationDates.slice(rttDaysUsed);
      const rttDates = vacationDates.slice(0, rttDaysUsed);

      seenRanges.add(signature);
      results.push({
        startDate: stripTime(new Date(start)),
        endDate: stripTime(new Date(end)),
        vacationDaysUsed: vacationDates.length,
        paidLeaveDaysUsed,
        rttDaysUsed,
        totalDaysOff,
        relatedHoliday: includedHolidays[0],
        includedHolidays,
        vacationDates,
        paidLeaveDates,
        rttDates,
        bridgeDates,
        weekendDates,
        schoolHolidayDates,
        holidayDate: includedHolidays[0].date,
        worthScore,
        rankingScore,
      });
    }
  }

  return results.sort((left, right) => {
    const rankingCompare = right.rankingScore - left.rankingScore;
    if (rankingCompare !== 0) {
      return rankingCompare;
    }
    const totalDaysCompare = right.totalDaysOff - left.totalDaysOff;
    if (totalDaysCompare !== 0) {
      return totalDaysCompare;
    }
    return left.startDate.getTime() - right.startDate.getTime();
  });
}

function containsSchoolHoliday(period: SchoolHolidayPeriod, date: Date): boolean {
  const normalizedDate = stripTime(date).getTime();
  return (
    normalizedDate >= stripTime(period.startDate).getTime() &&
    normalizedDate < stripTime(period.endDate).getTime()
  );
}

function periodsOverlap(left: BestVacationPeriod, right: BestVacationPeriod): boolean {
  return (
    stripTime(left.startDate).getTime() <= stripTime(right.endDate).getTime() &&
    stripTime(right.startDate).getTime() <= stripTime(left.endDate).getTime()
  );
}

function scoreAnnualCandidate(
  period: BestVacationPeriod,
  strategy: AnnualPlanningStrategy,
): number {
  switch (strategy) {
    case "family":
      return period.rankingScore + period.schoolHolidayDates.length * 0.45 + period.totalDaysOff * 0.04;
    case "balanced":
      return period.rankingScore + period.totalDaysOff * 0.05 - period.paidLeaveDaysUsed * 0.02;
    case "max-efficiency":
    default:
      return period.rankingScore + period.totalDaysOff * 0.03;
  }
}

function computeRankingScore({
  worthScore,
  schoolHolidayOverlapDays,
  preference,
}: {
  worthScore: number;
  schoolHolidayOverlapDays: number;
  preference: SchoolHolidayPreference;
}): number {
  switch (preference) {
    case "favor":
      return worthScore + schoolHolidayOverlapDays * 0.2;
    case "avoid":
      return worthScore - schoolHolidayOverlapDays * 0.2;
    case "neutral":
    default:
      return worthScore;
  }
}

function clamp(value: number, min: number, max: number): number {
  return Math.min(Math.max(value, min), max);
}

class BundleScore {
  constructor(
    readonly totalDaysOff: number,
    readonly totalBookedDays: number,
    readonly periodCount: number,
    readonly schoolHolidayDays: number,
  ) {}

  static empty() {
    return new BundleScore(0, 0, 0, 0);
  }

  add(period: BestVacationPeriod) {
    return new BundleScore(
      this.totalDaysOff + period.totalDaysOff,
      this.totalBookedDays + period.vacationDaysUsed,
      this.periodCount + 1,
      this.schoolHolidayDays + period.schoolHolidayDates.length,
    );
  }

  isBetterThan(other: BundleScore) {
    if (this.totalDaysOff !== other.totalDaysOff) {
      return this.totalDaysOff > other.totalDaysOff;
    }
    if (this.totalBookedDays !== other.totalBookedDays) {
      return this.totalBookedDays < other.totalBookedDays;
    }
    return this.periodCount > other.periodCount;
  }
}
